pkg/adapter/adapter.go

```diff
@@ -31,17 +36,141 @@ type envConfig struct {
+type dataSource interface {
+	Watch(ctx context.Context, pipeline interface{}, opts ...*options.ChangeStreamOptions) (*mongo.ChangeStream, error)
+}
+
+// NewEnvConfig creates an empty environement variables configuration.
 func NewEnvConfig() adapter.EnvConfigAccessor {
 	return &envConfig{}
 }
 
 // NewAdapter creates an adapter to convert incoming MongoDb changes events to CloudEvents and
-// then sends them to the specified Sink
+// then sends them to the specified Sink.
 func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client) adapter.Adapter {
-	fmt.Printf("Hey from mongodb adapter")
-	time.Sleep(600 * time.Second)
+	logger := logging.FromContext(ctx)
+	env := processed.(*envConfig)
+
+	return &mongoDbAdapter{
+		namespace:       env.Namespace,
+		ce:              ceClient,
+		source:          env.EventSource,
+		database:        env.Database,
+		collection:      env.Collection,
+		credentialspath: env.MongoDbCredentialsPath,
+		sinkurl:         env.SinkURL,
+		logger:          logger,
+	}
+}
+
+// Start connects to the database and creates the watch stream that will watch for dataSource changes.
+func (a *mongoDbAdapter) Start(ctx context.Context) error {
+	// Read the Credentials.
+	rawURI, err := ioutil.ReadFile(a.credentialspath + "/URI")
+	if err != nil {
+		return fmt.Errorf("Unable to get MongoDb URI field: secretPath %s : %w", a.credentialspath+"/URI", err)
+	}
+	URI := string(rawURI)
+	if URI == "" {
+		return fmt.Errorf("MongoDb URI field is an empty string: URI %s : %w", URI, err)
+	}
+
+	// Create new Client.
+	client, err := mongo.NewClient(options.Client().ApplyURI(URI))
+	if err != nil {
+		return fmt.Errorf("Error creating mongo client: %w", err)
+	}
+	// Get dataSource
+	dataSource := getDataSource(client, a.database, a.collection)
+
+	// Connect to Client.
+	err = client.Connect(ctx)
+	if err != nil {
+		return fmt.Errorf("Error connecting to database: %s", err)
+	}
+	defer client.Disconnect(ctx)
+
+	// Create a watch stream for either the database or collection.
+	stream, err := dataSource.Watch(ctx, mongo.Pipeline{})
+	if err != nil {
+		return fmt.Errorf("Error setting up changeStream: %s", err)
+	}
+	defer stream.Close(ctx)
+
+	// Watch and process changes.
+	a.processChanges(ctx, stream)
 	return nil
 }
+
+// Returns the appropriate dataSource: database or collection.
+func getDataSource(client *mongo.Client, db string, coll string) dataSource {
+	if coll != "" {
+		return client.Database(db).Collection(coll)
+	}
+	return client.Database(db)
+}
+
+// processChanges processes the new incoming change, creates a cloud event and sends it.
+func (a *mongoDbAdapter) processChanges(ctx context.Context, stream *mongo.ChangeStream) {
+	// For each new change recorded.
+	for stream.Next(ctx) {
+		var data bson.M
+		if err := stream.Decode(&data); err != nil {
+			a.logger.Desugar().Error("Error decoding the change stream", zap.Error(err))
```

Comment:
if there is an error here, you should. skip the iteration, otherwise it will break afterwards. Maybe add a continue, as you cannot return